compiler/rustc_resolve/src/build_reduced_graph.rs

```diff
@@ -913,14 +914,21 @@ impl<'a, 'b, 'tcx> BuildReducedGraphVisitor<'a, 'b, 'tcx> {
             let entry = self
                 .r
                 .extern_prelude
-                .entry(ident.normalize_to_macros_2_0())
+                .entry(ident)
                 .or_insert(ExternPreludeEntry { binding: None, introduced_by_item: true });
-            // Binding from `extern crate` item in source code can replace
-            // a binding from `--extern` on command line here.
-            entry.binding = Some(imported_binding);
             if orig_name.is_some() {
                 entry.introduced_by_item = true;
             }
+            // Binding from `extern crate` item in source code can replace
+            // a binding from `--extern` on command line here.
+            if ident.name != kw::Underscore && entry.binding.is_some() {
```

Comment:
```suggestion
            if ident.name != kw::Underscore && entry.introduced_by_item {
```
`entry.binding.is_some()` can be true for crates passed with `--extern my_crate` and used with `my_crate::something` in the source. 
We only need to skip overwriting for crates previously introduced by `extern crate` items.

Reply:
So, basically, the logic as I see it:
```
if !entry.introduced_by_item {
  entry.binding = Some(imported_binding);
} else if ident != kw::Underscore {
  span_delayed_bug();
}
```

Reply:
The binding is never set to `extern crate self as core;` in that case?
Looks like a bug.

Reply:
`entry.binding = Some(imported_binding)` should be done even if `entry.binding.is_some()` if the binding points to an `--extern` option (such bindings are lazily created in `fn extern_prelude_get`) and not to an `extern crate` item.

So, updating my condition:
```
if entry.binding.is_none() || !entry.introduced_by_item {
  entry.binding = Some(imported_binding);
} else if ident != kw::Underscore {
  span_delayed_bug();
}
```

Reply:
Changing the meaning of `introduced_by_item` will break the logic in which it is used.

Maybe check the kind of `entry.binding`?
It must be an import for `extern crate`s and not an import for `--extern`s.

Reply:
> Maybe check the kind of entry.binding?

I think the above case is not related to `entry.binding` because the issue is caused by `!entry.introduced_by_item`.

And the changes to `introduced_by_item` I envisioned are about that:

```rs
let old_introduced_by_item = entry.introduced_by_item;
if orig_name.is_some() || item.name == "core || item.name == "std" {
    entry.introduced_by_item = true;
}
// then
if entry.binding.is_none() || !old_introduced_by_item {
  entry.binding = Some(imported_binding);
} else if ident != kw::Underscore {
  span_delayed_bug();
}
```

Reply:
I'm starting to lose the track.
By checking the import kind I meant
```rust
if entry.binding.is_none() || !entry.binding.is_import() {
  entry.binding = Some(imported_binding);
} else if ident != kw::Underscore {
  span_delayed_bug();
}
```
The we either set a new binding, or overwrite binding from `--extern` or builtin `core`/`std` but not any `extern crate`, which seems to cover all the cases.
(The meaning of `entry.introduced_by_item` stays unchanged.)

Reply:
I think I get what you mean: we can use `!entry.is_import()` as the test condition

Reply:
Ah, great, there's even a method for it.